### PR TITLE
new source widget

### PR DESCRIPTION
**Pull request by mcoughlin**
This PR adds a new source widget to the frontend.

### CODE REVIEW

**Review by profjsb — CHANGES_REQUESTED** [review]
1) Entry for dec and submit is spilling outside the widget for all viewport sizes:
Kapture 2022-03-10 at 11 26 44

Consider making more compact the entries.

2) Missing hamburger icon to move widget elsewhere on the dashboard. See how other widgets are implemented.

3) This widget should be checking the object ID and RA/DEC for potential conflicts. One idea is to use a popup...

4) Popup confirm and take user to the newly formed source as such:
Kapture 2022-03-10 at 12 44 52

**Comment by mcoughlin** [comment]
@profjsb did you mock this up in that gif?? I am having some difficulty with the handling of the checking for existing sources.

**Comment by mcoughlin** [comment]
@profjsb I wrote a custom endpoint to check for existing sources and push those messages back. I hope that suffices.

File: skyportal/handlers/api/source_exists.py [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+            description: Radius for spatial filtering if ra & dec are provided (in decimal degrees)
+        """
+
+        ra = self.get_query_argument('ra', None)
+        dec = self.get_query_argument('dec', None)
+        radius = self.get_query_argument('radius', None)
+
+        if obj_id is not None:
+            s = Obj.get_if_accessible_by(obj_id, self.current_user)
+            print(obj_id)
```
Comment (guynir42) [inline skyportal/handlers/api/source_exists.py:57]:
```suggestion
```

File: skyportal/handlers/api/source_exists.py [hunk 2]
```diff
@@ -0,0 +1,87 @@
...
+            if len(objs) == 1:
+                return self.success(
+                    f"A source at that location already exists: {objs[0].id}."
+                )
+            elif len(objs) > 1:
+                return self.success(
+                    f"Sources at that location already exist: {','.join([obj.id for obj in objs])}."
+                )
+
+        return self.success(False)
```
Comment (guynir42) [inline skyportal/handlers/api/source_exists.py:87]: Don't we already have this functionality in the source handler using the HEAD request?

File: static/js/components/NewSource.jsx [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+            <GroupShareSelect
+              groupList={allGroups}
+              setGroupIDs={setSelectedGroupIds}
+              groupIDs={selectedGroupIds}
+            />
+            <Form
+              schema={sourceFormSchema}
+              onSubmit={handleSubmit}
+              // eslint-disable-next-line react/jsx-no-bind
+              validate={validate}
```
Comment (guynir42) [inline static/js/components/NewSource.jsx:87]: why not `liveValidate`?

**Review by guynir42 — COMMENTED** [review]
My main concern is that you may be able to get the same functionality using the source handler HEAD request.

File: skyportal/handlers/api/source_exists.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/handlers/api/source_exists.py:87]: We also need to check the nearby RA and Dec per Josh's request.

**Comment by guynir42** [comment]
Also, when playing around with it I noticed a couple of small issues.
1) the first field "Share Data With": the text is too big for the box. 
2) When you add a source, it complains that it cannot find the source. Also the photometry and spectroscopy spinners go on forever. I think these might be related. Maybe we should make sure that a source without photometry/spectroscopy is handled gracefully, and also the spinners should turn into "no photometry" or "no spectroscopy" messages.

File: skyportal/handlers/api/source_exists.py (hunk 2 above)
Comment (guynir42) [reply skyportal/handlers/api/source_exists.py:86]: The source handler doesn't do a cone search?

**Comment by mcoughlin** [comment]
@guynir42 Yeah for now I am throwing an "error" for the source catch. Not sure the best way to catch that. The spinners end for me though... what are you seeing?

File: static/js/components/NewSource.jsx (hunk 3 above)
Comment (mcoughlin) [reply static/js/components/NewSource.jsx:87]: It's a little annoying the way it whines immediately once you input a name. maybe see if you prefer with or without it?

File: skyportal/handlers/api/source_exists.py (hunk 2 above)
Comment (mcoughlin) [reply skyportal/handlers/api/source_exists.py:86]: The head handler doesn't.

**Comment by guynir42** [comment]
Spinners seem to remain forever, I don't know why. 

I urge you to consider adding the cone search to sources HEAD, we have too many handlers as it is. 

I think there must be a simple way to return a source or nothing without popping up an error.

**Comment by mcoughlin** [comment]
@guynir42 HEAD doesn't seem to be an API call available in the duck.

**Comment by guynir42** [comment]
@mcoughlin that's weird. I think this functionality is used for candidates, but maybe it is only called internally. 
Did you find an alternative to returning an error if the source already exists / doesn't exist?

**Comment by mcoughlin** [comment]
@guynir42 Where do you actually see this displayed?

**Comment by mcoughlin** [comment]
@guynir42 maybe we hop on zoom briefly. I don't quite see it but I suspect just I don't know where to see.

**Comment by guynir42** [comment]
sure let's do it around 9

**Comment by mcoughlin** [comment]
@guynir42 can you see if this fix does it for you?

**Comment by guynir42** [comment]
The photometry and spectroscopy look good. The "Source not found" error still pops up.

**Comment by mcoughlin** [comment]
I am not sure how to block this one without screwing the normal behavior up... @guynir42

**Comment by guynir42** [comment]
Is the error produced by the photometry handler?

**Comment by mcoughlin** [comment]
Source handler.

**Comment by guynir42** [comment]
Is it produced in any situation where there is no photometry?

**Comment by mcoughlin** [comment]
I think so?

**Comment by guynir42** [comment]
That doesn't sound like the right behavior to me. Try to see if there's a better criteria for returning such an error. Otherwise maybe we can live without this error? In what situations do we need that to return an error anyway?

**Comment by mcoughlin** [comment]
@guynir42 can i ask if it triggers for you if we remove the push.history call in the NewSource handler?

**Comment by guynir42** [comment]
It is unrelated to the new source. It pops up whenever I refresh the source page (presumably for sources without photometry).

**Comment by mcoughlin** [comment]
And goes away if you add photometry? I can't seem to reproduce this.

**Comment by mcoughlin** [comment]
@jadalilleboe what do you see?

**Comment by jadalilleboe** [comment]
I don't see a source not found error.

**Comment by guynir42** [comment]
hmmm let me dig into this. I thought you were seeing this error as well.

**Comment by guynir42** [comment]
@mcoughlin it seems to only pop up for sources with a space in the name. The handlers that return the error are photometry and spectroscopy. This probably has to do with the parsing of the source ID. I'd say this is not really relevant to this PR.

**Comment by guynir42** [comment]
I did notice, that when there are errors on the submit new source form, it drives the rest of the input fields off the bottom of the component.

**Comment by mcoughlin** [comment]
@guynir42 do you think the benefit of live validate offsets the UI or should we just turn it back off?

**Comment by guynir42** [comment]
If the behavior was ok without the live validate, I think we can live without it (there aren't a lot of inputs and the errors are fairly obvious).

**Comment by mcoughlin** [comment]
@guynir42 Ok removing for now.

**Review by guynir42 — APPROVED** [review]

**Pull request merged**